### PR TITLE
Thread safe retry for s3.

### PR DESCRIPTION
**Pull request by MikhailBurdukov**
Switched `retry` decorator from tenacity  to  own `retry_exponential` one.

### CODE REVIEW

File: ch_backup/storage/engine/s3/s3_retry.py [hunk 1]
```diff
@@ -57,3 +62,93 @@ def wrapper(self: "S3StorageEngine", *args: Any, **kwargs: Any) -> RT:
...
+
+        def calculate_sleep_time(self, retry_count):
+            """
+            Calculate time to sleep.
+            """
+
+            high_value = (2**retry_count) * multiplier
+            return uniform(0, min(high_value, max_interval))  # nosec
+
+        def call(self, fn, *args, **kwargs):
```
Comment (aalexfvk) [inline ch_backup/storage/engine/s3/s3_retry.py:109]: Consider overriding  `__call__` method

File: ch_backup/storage/engine/s3/s3_retry.py [hunk 2]
```diff
@@ -57,3 +62,93 @@ def wrapper(self: "S3StorageEngine", *args: Any, **kwargs: Any) -> RT:
...
+                            time_to_sleep,
+                            attempts,
+                            exc,
+                        )
+
+                    sleep_function(self.calculate_sleep_time(retry_count))
+
+    def wrap(f):
+        @wraps(f)
+        def wrapped_f(*args, **kw):
```
Comment (aalexfvk) [inline ch_backup/storage/engine/s3/s3_retry.py:149]: `kwargs` is more common name

File: tests/unit/utils.py [hunk 3]
```diff
@@ -177,3 +177,14 @@ def assert_equal(actual, expected):
         assert (
             False
         ), f"\n   expected: {expected}\n    but was: {actual}\ndifferences:\n{diff.pretty()}"
+
+
+class TimeMocker:
```
Comment (aalexfvk) [inline tests/unit/utils.py:182]: Let's move it to the separate module `time_mocker.py`

**Review by aalexfvk — CHANGES_REQUESTED** [review]

File: ch_backup/storage/engine/s3/s3_retry.py [hunk 4]
```diff
@@ -57,3 +61,92 @@ def wrapper(self: "S3StorageEngine", *args: Any, **kwargs: Any) -> RT:
...
+        self.sleep_function = sleep_function
+        self.verbose = verbose
+
+    def calculate_sleep_time(self, attempt):
+        """
+        Calculate time to sleep.
+        """
+
+        high_value = (2**attempt) * self.multiplier
+        return uniform(0, min(high_value, self.max_interval))  # nosec
```
Comment (aalexfvk) [inline ch_backup/storage/engine/s3/s3_retry.py:118]: Nevertheless, I think fixed interval backoff is more preferable here. Pay attention on [this](https://github.com/jd/tenacity/blob/310058274ed22a345e9c3917c97b4afd6363d5a5/tenacity/wait.py#L186) suggestion from tenacity. Clickhouse itself uses exponential fixed backoff intervals when requsting S3.

File: tests/unit/utils.py (hunk 3 above)
Comment (aalexfvk) [reply tests/unit/utils.py:182]: Forget to delete ?

File: ch_backup/storage/engine/s3/s3_retry.py [hunk 5]
```diff
@@ -57,3 +61,92 @@ def wrapper(self: "S3StorageEngine", *args: Any, **kwargs: Any) -> RT:
...
+
+    def __call__(self, fn, *args, **kwargs):
```
Comment (aalexfvk) [inline ch_backup/storage/engine/s3/s3_retry.py:120]: Write comment please that an instance of this class can be called from different threads and should be thread safe. For example, here we don't update state of the class (instance variables) and use only local ones.

**Review by aalexfvk — APPROVED** [review]

**Comment by Alex-Burmak** [comment]
Test failures for `latest` ClickHouse versions don't relate to the changes from this PR. It relates to https://github.com/yandex/ch-backup/issues/84.

 Merging the PR.

**Pull request merged**